### DISCUSSION TITLE
RewardTrainer Producing Same Reward Score During and After Training

### DISCUSSION THREAD

**Discussion by deepakpandita57** [huggingface/trl#2265] [Q&A]
I'm currently working on a project using the TRL library for training a reward model with the RewardTrainer following the example from the trl repo. However, I've encountered an issue where the model consistently produces the same reward score during training and after evaluation. Here are some relevant details about my setup:

Model and Setup:
* Model Name: meta-llama/Meta-Llama-3-8B-Instruct
* Training Epochs: 3
* Batch Size: 8 (for both training and evaluation)
* Learning Rate: 1e-5
* Max Length: 1024
* Dataset: CarperAI/openai_summarize_comparisons
* Initially I started with a sample of 1,000 training examples and gradually increased to 20,000, validation on 100 examples just to test

Observations:
* The reward scores are identical for the training and validation data points, indicating no learning is occurring.
* The metrics during training do not show any improvement over epochs.
* I’ve monitored the training and validation loss, but they remain stagnant throughout the process. Sometimes with different settings with/without PEFT, the training loss does change and the trainer.evaluate() shows different logits for chosen and rejected text indicating that the model is learning. However, it outputs the same score for chosen vs rejected text and the score is sometimes different for different data points.

Questions:
1. What could be the possible reasons for the model not learning effectively?
2. Are there specific hyperparameters or configurations in the TRL library that I should consider adjusting?
3. Could there be any issues with how I’m implementing the reward model or the training process?

I understand you might need further details to point out the specific issue, I'm willing to provide the necessary details.

Any insights or suggestions would be greatly appreciated!
Thank you!

**Comment by jijivski** [comment]
Same question, any update? 
I tried to reproduce RLHF-Reward-Modeling's BT model using Meta-Llama-3-8B-Instruct.
```
cd RLHF-Reward-Modeling/ 
CUDA_VISIBLE_DEVICES=0,1,2,3 accelerate launch --main_process_port 29502 ./bradley-terry-rm/gemma_2B_rm_orig.py --deepspeed deepspeed_configs/deepspeed_2.json --model_name Meta-Llama-3-8B-Instruct --max_length 4096 --train_set_path hendrydong/preference_700K --per_device_train_batch_size 2 --per_device_eval_batch_size 2 --gradient_accumulation_steps 4 --learning_rate 2e-05 --weight_decay 0.0005  --output_path bt_models/Meta-Llama-3-8B-Instruct/
```
but models like gemma-2b-it and Qwen2.5-0.5B-Instruct are right.

**Comment by jijivski** [comment]
I transfered this problem to  #https://github.com/RLHFlow/RLHF-Reward-Modeling/issues/49, and got help from the author, 
see this code: https://github.com/RLHFlow/RLHF-Reward-Modeling/blob/main/bradley-terry-rm/llama3_8B_rm.py
